tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,163 @@
+use std::sync::atomic::{AtomicU16, AtomicU8, Ordering};
+
+use crate::Client;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+struct ___tracy_gpu_new_context_data {
+    gpu_time: i64,
+    period: f32,
+    context: u8,
+    flags: u8,
+    type_: u8,
+}
+
+#[repr(u8)]
+///
+pub enum GpuContextType {
+    ///
+    Invalid,
+    ///
+    OpenGL,
+    ///
+    Vulkan,
+    ///
+    OpenCL,
+    ///
+    Direct3D12,
+    ///
+    Direct3D11,
+}
+
+///
+pub struct GpuContext {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+}
+#[cfg(feature = "enable")]
+static GPU_CONTEXT_INDEX: AtomicU8 = AtomicU8::new(0);
+
+///
+pub struct GpuSpan {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    start_query_id: u16,
+    #[cfg(feature = "enable")]
+    end_query_id: Option<u16>,
+}
+#[cfg(feature = "enable")]
+static GPU_SPAN_INDEX: AtomicU16 = AtomicU16::new(0);
+
+impl Client {
+    ///
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> GpuContext {
+        #[cfg(feature = "enable")]
+        unsafe {
+            let context = GPU_CONTEXT_INDEX.fetch_add(1, Ordering::Relaxed);
+
+            sys::___tracy_emit_gpu_new_context(std::mem::transmute(
+                ___tracy_gpu_new_context_data {
+                    gpu_time: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
```

Comment:
If we're relying on particular values for this enum, I'd prefer to declare them explicitly in the enum definition.

Reply:
Still needs to be done.

Reply:
Explicit values specified.